### PR TITLE
ARROW-6930: [Java] Create utility class for populating vector values used for test purpose only

### PR DESCRIPTION
**Pull request by tianchen92**
There is a lot of verbosity in the construction of vectors for testing purposes (multiple lines of setSafe(...) or set(...). This change adds a utility to populate vector values in a concise way for test purposes only. The utility class TestValueVectorPopulator is located in java/vector/src/test/java/org/apache/arrow/vector/testing and can be used for testing by modules other than arrow-vector by adding the following test dependency:

```
<dependency>
  <groupId>org.apache.arrow</groupId>
  <artifactId>arrow-vector</artifactId>
  <version>${project.version}</version>
  <classifier>tests</classifier>
  <type>test-jar</type>
  <scope>test</scope>
</dependency>
```

Usage would be something like:
```
try (IntVector vector = new IntVector(“vector”, allocator)) {
    ValueVectorDataPopulator.setVector(vector, 1, 2, null, 4, 5);
    output = doSomethingWith(input); 
    assertThat(output).isEqualTo(expected); 
}
```

### CODE REVIEW

**Comment by tianchen92** [comment]
cc @emkornfield

File: java/vector/src/main/java/org/apache/arrow/vector/BitVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVector.java]: I think we should use object types here and make it variable length arguments so it can be used as:

create("vectorName", allocator, true, false, null) )i.e. no need to clutter with creating a new array when using the method and the contact is clearer (same goes for other methods)

**Comment by emkornfield** [comment]
@BryanCutler @praveenbingo I'd like to get your thoughts on an API that makes sense

File: java/vector/src/main/java/org/apache/arrow/vector/BitVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/BitVector.java]: ok, sounds reasonable. Also there is a special case IntervalDayVector takes two values(days and milliSeconds) which is not suitable for this pattern, so keep this vector with arrays?

**Comment by TheNeuralBit** [comment]
This API seems reasonable to me, especially if it's intended primarily as a convenience for testing.

**Comment by jacques-n** [comment]
Couple thoughts:

- This is really a form of builder. Let's keep them out of the core vector objects. They have no need to be inside those objects. (This also keeps the core objects cleaner). 
- We really need to decide whether these are generally useful or for testing. If generally useful, let's make them for general purpose. If only for testing purposes, let's put the impl in a test package, not in the test directories, not the main directories
- If we want to link so people can use from the existing vector classes (if we decide these are generally useful), let's keep it to something akin to BigIntVector.newBuilder() with the Builder object outside (but that is only if we think they are not for testing).
- Let's support building with both boxed and unboxed types for primitives.

In general, let's have these conversations before you go and write a couple hundred lines of code. Will waste less of your time.

**Comment by emkornfield** [comment]
@jacques-n 

> We really need to decide whether these are generally useful or for testing.

My impression is we wouldn't want to encourage usage of these for general code, and there is already the Writer* classes for adding values?  Do you think there is a use-case for general purpose methods/builder class?  Maybe a discussion the ML.

>  If generally useful, let's make them for general purpose. If only for testing purposes, let's put the impl in a test package, not in the test directories, not the main directories

Can you clarify do you mean? A separate maven package or a testing java package under main? Or exporting a test jar from the arrow-vector class?  Possibly, relevant [SO question](https://stackoverflow.com/questions/174560/sharing-test-code-in-maven#174670) 

> If we want to link so people can use from the existing vector classes (if we decide these are generally useful), let's keep it to something akin to BigIntVector.newBuilder() with the Builder object outside (but that is only if we think they are not for testing).

I think it is fine to have a separate builder if we think that it is valuable but I think that is orthogonal to simple static factory methods. 
A good example of an API that has both is [Guava's Immutable Structures](https://guava.dev/releases/20.0/api/docs/com/google/common/collect/ImmutableMap.html).  Both static factories and builders have their place, and it make it convenient if they are both on the top level class.

> Let's support building with both boxed and unboxed types for primitives.

I think this only makes sense we make these general purpose.  Was there a use-case you were thinking of in particular for the testing scenario?

File: java/vector/src/main/java/org/apache/arrow/vector/TinyIntVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/TinyIntVector.java]: style: Byte... (no space)

**Comment by jacques-n** [comment]
> My impression is we wouldn't want to encourage usage of these for general code, and there is already the Writer* classes for adding values? Do you think there is a use-case for general purpose methods/builder class? Maybe a discussion the ML.

My comments were: if for test only, they shouldn't bleed into the non-test code (straight on vectors).

> Can you clarify do you mean? A separate maven package or a testing java package under main? Or exporting a test jar from the arrow-vector class?
I just meant put it in src/test/java, not src/main/java. We can depend on the test impls in other module via a maven test dependency if we want to use them outside vector module. We do that frequently at Dremio to use common test code without leaking it into main production code.


> A good example of an API that has both is [Guava's Immutable Structures](https://guava.dev/releases/20.0/api/docs/com/google/common/collect/ImmutableMap.html). Both static factories and builders have their place, and it make it convenient if they are both on the top level class.

To me that is the crux of the question. If for common use, linking a static method off of the vector classes make sense. If not, let's not. In the immutable case, the links are off the main class because the interfaces are for production use.


> I think this only makes sense we make these general purpose. Was there a use-case you were thinking of in particular for the testing scenario?

Yes, just general purpose. For test, agree that not important.

My stance is basically:
- if test only, avoid polluting non-test interfaces/codebase.
- if not test, keep separate and just do a single link a'la builder. Since we use builder metaphor in C++ code, it might be nice to expose same concepts here. Also, if done via builder, maybe we could reuse/expand the ComplexWriter subtrees, etc rather than implementing an separate impl (just brainstorming here, not sure if this makes sense).

**Comment by emkornfield** [comment]
Thanks @jacques-n sounds reasonable.  I'm leaning potentially towards test only for now and we can maybe discuss the general purpose usage on the mailing list.  There are a few other committers I tagged on the PR, so i'd like to hear their thoughts as well as  @tianchen92 before we do any refactoring

**Comment by jacques-n** [comment]
Sounds good. Thanks for helping facilitate this work @emkornfield! Appreciate the proposal @tianchen92.

**Comment by tianchen92** [comment]
@jacques-n Thanks for your suggestion, I'll hold off this until we reach consistent. Hmm, this was proposed by @emkornfield, thanks Micah!

**Comment by BryanCutler** [comment]
I'm kind of neutral on this change. If this ends up not just test-only, then I agree with @jacques-n and we should design a more comprehensive builder and reuse the ComplexWriter. It's not a good idea to have multiple ways to write to a vector that are similar.

If this is just for test-only, I think trying to squeeze the test values in the same line to create the vector, that is already in a `try` statement, is too much and gets difficult to read for anything more than a couple of values. Most of the verbosity is in the calls to `allocateNew`, `setValueCount` and I'd rather see something like this that would make it easier to set some values at once.

```java
// Create vector like before
try (final IntVector v = new IntVector("intVector", allocator)) {

  // calls `allocateNew`, `setValueCount`, `setSafe`
  setVector(v, 1, 2);
  ...
}

**Comment by emkornfield** [comment]
OK based on the conversation lets make these test only.  Maybe a class called VectorValueDataPopulator with static methods as @BryanCutler described above?

**Comment by tianchen92** [comment]
> OK based on the conversation lets make these test only. Maybe a class called VectorValueDataPopulator with static methods as @BryanCutler described above?

LGTM, @jacques-n do you think this is ok? if there's no objection, I would refactor this way later today.

**Comment by jacques-n** [comment]
Test only? Sounds good to me.

**Comment by tianchen92** [comment]
Thanks @emkornfield @jacques-n @BryanCutler for your efforts, I have updated this PR, please take a look again.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: it would be nice to have a string override for this.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: this is fine for now, in the future we might want to add support for the next size up to support the full range of values easily (same goes for Uint4 and Uint8

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: technically, I think setValueCount should be called after setting the values?

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: no capital S in Stamp.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: javadoc for in general would be helpful. for all methods (it doesn't need to cover all parameters for obvious ones).  For timestamp where the value has specific semantics it would be good to document what it represents (e.g. nanonseconds since the unix epoch)

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: private constructor for now.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: we should have explicit tests for this class.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: can you demonstrate the usage of this class in at least one instance in a different package tests?  I want to make sure it is reusable.

**Review by emkornfield — CHANGES_REQUESTED** [review]
See comments I left, I think the API is reasonable.  @BryanCutler @TheNeuralBit any additional thoughts?

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: Now this class was located in vector test package, should we move it into vector main package?

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: Hmm, this is the class name?

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: should test for all types?

Comment (emkornfield) [reply java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: Yes please.

File: java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java
Comment (emkornfield) [reply java/vector/src/test/java/org/apache/arrow/vector/ValueVectorDataPopulator.java]: I think th easiest thing to do would be to move it to main, under a new java package "org.apache.arrow.vector.testing".  Another alternative (which I don't think would work) is create an entirely new mvn submodule and put it in main there.

I think @jacques-n might have concerns around this, so lets wait for him to chime in (I think I posted earlier about a possible way of exposing the testing jar to other packages but it seemed hacky).

**Comment by tianchen92** [comment]
> See comments I left, I think the API is reasonable. @BryanCutler @TheNeuralBit any additional thoughts?

Thanks, PR updated.

File: java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: should this go under src/test?

File: java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Would it make sense to use a generic function that would parametrize the vector and value types to remove the duplication? Since it's for testing only, not too big of deal..

**Review by BryanCutler — COMMENTED** [review]
Looks pretty good, I just had a couple of questions

**Comment by BryanCutler** [comment]
Also, make sure to update the PR title and desc

File: java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: thanks, previous version was under src/test, to make it reusable for other module tests, @emkornfield suggested moving it to main, and we are waiting for @jacques-n to confirm. see https://github.com/apache/arrow/pull/5693#discussion_r338893344

File: java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Hmm, seems set API could not use generic type? and if so comments for how to set values like date type vectors would be removed either.

Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Ok, just wondering if it was possible to reduce the duplication some, but this is fine

File: java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Ok, I'll let @jacques-n make the call, but my preference would be to not include this in the main artifact. There should be some other way for additional modules to use this in tests, if needed.

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: I think the steps described in https://cleantestcode.wordpress.com/2014/04/13/reusable-test-utility-code-and-maven/ make sense to keep it out of main.  (I prefer the naming "testing" to testutils)

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: I followed the steps in the link you provided, but failed to work...

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: can you provide more context on didn't work?

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: mvn clean install failed which says could not found this class. I didn't keep the original message.

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Hmm, OK, based on https://maven.apache.org/plugins/maven-jar-plugin/examples/create-test-jar.html it seems like we really should create a new module for this code (or you can try to follow the first example, but I think that is the same as the one I linked above).

Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java]: Hmm, I tried to follow the first example and it worked :), now the utility class is under test package and I add a test case in Avro module to make sure it is reusable.
Please see if it looks reasonable @emkornfield

File: java/vector/src/test/java/org/apache/arrow/vector/TestDictionaryVector.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/TestDictionaryVector.java]: we should either include the character set here of make use just make  the use of "aa" possible.

File: java/vector/src/test/java/org/apache/arrow/vector/testing/TestValueVectorPopulator.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/testing/TestValueVectorPopulator.java]: spacing issue here.

**Review by emkornfield — CHANGES_REQUESTED** [review]
two minor comments, otherwise looks ok to me.

It would be nice to have a real unit test in a separate package rather then one just meant for testing this change.

@BryanCutler does this look okay to you?

File: java/vector/src/test/java/org/apache/arrow/vector/TestDictionaryVector.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/TestDictionaryVector.java]: Thanks, fixed.

File: java/vector/src/test/java/org/apache/arrow/vector/testing/TestValueVectorPopulator.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/testing/TestValueVectorPopulator.java]: fixed.

File: java/vector/pom.xml
Comment (BryanCutler) [inline java/vector/pom.xml]: Just to clarify - so this packages vector tests in a jar that other modules can depend on?

Comment (BryanCutler) [reply java/vector/pom.xml]: Actually, I think this is already declared in the root pom and these *-tests.jars already exist. So this isn't needed right?

Comment (tianchen92) [reply java/vector/pom.xml]: Right, removed.

File: java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java
Comment (BryanCutler) [inline java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java]: It would be better if this was an actually useful test for Avro, but at least could you add a comment that this is meant to only test out the dependence on the vector tests artifact? Otherwise, it's a little confusion why this test is here at all.

**Review by BryanCutler — COMMENTED** [review]
LGTM, just one minor comment

**Comment by BryanCutler** [comment]
Again, please update the PR title/description

File: java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java
Comment (tianchen92) [reply java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java]: Thanks, I added a comment to clarify this.

Comment (BryanCutler) [reply java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java]: Sorry, this is kind of bugging me having a pointless test here. How about moving the dependency to arrow.tools and change this test to use it? https://github.com/apache/arrow/blob/master/java/tools/src/test/java/org/apache/arrow/tools/EchoServerTest.java#L157

**Comment by BryanCutler** [comment]
By updating the description, I mean this https://github.com/apache/arrow/pull/5693#issue-329710750 that will end up in the commit message.

File: java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java
Comment (tianchen92) [reply java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowTest.java]: ok, fixed.

**Comment by tianchen92** [comment]
> By updating the description, I mean this [#5693 (comment)](https://github.com/apache/arrow/pull/5693#issue-329710750) that will end up in the commit message.

Thanks for clarify, fixed.

**Review by BryanCutler — APPROVED** [review]
Thanks for updating, LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @tianchen92 !

**Pull request closed without merging**
